plugins/Vibed/Vibed.cpp

```diff
@@ -255,425 +215,326 @@ void Vibed::loadSettings( const QDomElement & _this )
 			// - save and restore selected wave-shape-button
 		}
 	}
-	
-//	update();
+	// update();
 }
 
-
-
-
 QString Vibed::nodeName() const
 {
-	return( vibedstrings_plugin_descriptor.name );
+	return vibedstrings_plugin_descriptor.name;
 }
 
-
-
-
-void Vibed::playNote( NotePlayHandle * _n, sampleFrame * _working_buffer )
+void Vibed::playNote(NotePlayHandle* n, sampleFrame* workingBuffer)
 {
-	if ( _n->totalFramesPlayed() == 0 || _n->m_pluginData == nullptr )
+	if (/*n->totalFramesPlayed() == 0 ||*/ n->m_pluginData == nullptr)
```

Comment:
All the other instrument plugins use the same condition before creating the object to store in `n->m_pluginData`, but it seems to me like a subtle bug waiting to happen if `n->totalFramesPlayed()` is ever zero while `n->m_pluginData` is non-null, you know? Seems like we should only be checking one not both, but I wasn't completely sure about it which is why I just commented it out.

Reply:
> it seems to me like a subtle bug waiting to happen if `n->totalFramesPlayed()` is ever zero while `n->m_pluginData` is non-null

If I remember correctly, that might happen when the tempo is heavily automated.

Reply:
I just did a more thorough investigation into this.

**TLDR**: Yes, a heavily automated tempo breaks it and causes a memory leak. This memory leak affects almost all instrument plugins.

**Long version**:
In most instrument plugins' `playNote` routines, if the condition `m_totalFramesPlayed == 0 && m_pluginData != nullptr` ever occurs, the `m_pluginData` pointer would be overwritten with a pointer to new dynamically allocated memory and a memory leak would occur. For said condition to occur, `m_instrumentTrack->playNote(...)` would need to have been called previously (which is the only way for `m_pluginData != nullptr` to happen), and `m_totalFramesPlayed` would need to have been set to zero sometime after that. So the question becomes "Is it possible for `m_totalFramesPlayed` be set to zero prior to calling `m_instrumentTrack->playNote(...)` again?" (The answer is yes)

These are the only three places I can find where the value of `m_totalFramesPlayed` changes:
**(1)** The NotePlayHandle constructor initializes it to zero.
**(2)** `NotePlayHandle::play(...)` increases it by `framesThisPeriod` at the end before returning.
**(3)** `NotePlayHandle::resize(...)` sets it to zero when called with `m_origFrames == 0` or `m_origTempo == 0`.

Possibility (1) can be ignored since the constructor is only called once at the start before `m_instrumentTrack->playNote(...)` is ever called, and it sets `m_pluginData` to nullptr, so it cannot cause the condition `m_totalFramesPlayed == 0 && m_pluginData != nullptr`.

As for possibility (2), it looks like `framesThisPeriod` is always >0 when set, so `m_totalFramesPlayed` should always increase at the end of `NotePlayHandle::play(...)`. And if `m_instrumentTrack->playNote(...)` is called, it will create the condition `m_pluginData != nullptr`, but it looks like it's guaranteed to continue on to the end of `NotePlayHandle::play(...)` where `m_totalFramesPlayed` is then increased (assuming no exceptions are thrown). This all means that `NotePlayHandle::play(...)` cannot set `m_totalFramesPlayed` to zero, always increases it before `m_instrumentTrack->playNote(...)` is called again, and therefore cannot create the condition where `m_totalFramesPlayed == 0 && m_pluginData != nullptr`. So possibility (2) can also be ignored.

This leaves the last possibility that I can see, which is (3). If `NotePlayHandle::resize(...)` were to be called sometime after the first time `m_instrumentTrack->playNote(...)` is called (so that `m_pluginData != nullptr`), it could set `m_totalFramesPlayed` to zero from some non-zero value it previously was. This would happen when `m_origFrames == 0` or `m_origTempo == 0`. I assume `m_origTempo` is never 0, so I'll ignore it. `m_origFrames` is only set to 0 if NotePlayHandle is constructed with the `_frames` parameter ==0. `NotePlayHandle::resize(...)` is only called if `Song::setTempo()` is called, which tracks with what you said PhysSong.

All of this means that the situation where the condition `m_totalFramesPlayed == 0 && m_pluginData != nullptr` would occur - thus causing memory leaks in most instrument plugins - looks like this:
1) A NotePlayHandle is created with `_frames` == 0.
2) `NotePlayHandle::play(...)` is called, and then `m_instrumentTrack->playNote(...)` is called from that.
3) The song's tempo changes, which calls `NotePlayHandle::resize(...)` and sets `m_totalFramesPlayed` back to 0.
4) Step 2 happens again, causing a memory leak.

To test whether I could actually cause this to happen, I added the following assert to the Vibed `playNote` routine:

```c++
assert((n->m_pluginData == nullptr) == (n->totalFramesPlayed() == 0));
```

Then I looped a section of a track with lots of Vibed notes playing and heavy tempo automation. The assert soon failed, meaning that a memory leak would have occurred had it still been using the old `n->totalFramesPlayed() == 0 || n->m_pluginData == nullptr` or `n->totalFramesPlayed() == 0` condition like most other instrument plugins still currently use.

I propose that we fix those conditions in the other instrument plugins, and maybe add some asserts in the `NotePlayHandle` code to make it clear which conditions hold true, what assumptions are being made, help demystify it, etc. Does that sound okay @PhysSong?

Reply:
One question that remains is what the solution should look like for each `playNote` routine.
I'm thinking either this:

```c++
if (n->totalFramesPlayed() == 0 || n->m_pluginData == nullptr)
{
    if (n->m_pluginData == nullptr)
    {
        deleteNotePluginData(n);
    }
    
    // ...

    n->m_pluginData = /* new plugin data here */
}
```

Or this:

```c++
if (n->m_pluginData == nullptr)
{
    // ...
    n->m_pluginData = /* new plugin data here */
}
```
The first option basically resets the plugin data to the default state whenever `m_totalFramesPlayed` is set back to zero, which is how most plugins currently operate - intended or not - while still fixing the memory leak.

Reply:
The latter sounds better to me.